src/utils/tests.py

```diff
@@ -1103,3 +1104,37 @@ def test_mailgun_webhook(self):
+
+    @mock.patch('utils.orcid.get_orcid_record', return_value=all_fields)
+    def test_record_details_all(self, mock_record):
+        details = get_orcid_record_details("0000-0000-0000-0000")
+        self.assertEqual(details["orcid"], "0000-0000-0000-0000")
+        self.assertEqual(details["uri"], "http://sandbox.orcid.org/0000-0000-0000-0000")
+        self.assertEqual(details["first_name"], "cdleschol")
+        self.assertEqual(details["last_name"], "arship")
+        self.assertEqual(len(details["emails"]), 1)
+        self.assertEqual(details["emails"][0], "cdleschol@mailinator.com")
+        self.assertEqual(details["affiliation"], "California Digital Library")
+        self.assertEqual(details["country"], "US")
+
+    @mock.patch('utils.orcid.get_orcid_record', return_value=min_fields)
+    def test_record_details_min(self, mock_record):
+        details = get_orcid_record_details("0000-0000-0000-0000")
+        self.assertEqual(details["orcid"], "0000-0000-0000-0000")
+        self.assertEqual(details["uri"], "http://sandbox.orcid.org/0000-0000-0000-0000")
+        self.assertIsNone(details["first_name"])
+        self.assertIsNone(details["last_name"])
+        self.assertEqual(len(details["emails"]), 0)
+        self.assertIsNone(details["affiliation"])
+        self.assertIsNone(details["country"])
+
+    def test_redirect_uri(self):
+        press= helpers.create_press()
+        repo = helpers.create_repository(press, [], [])
+        self.assertEqual(build_redirect_uri(repo), "http://localhost/login/orcid/?action=login")
+        self.assertEqual(build_redirect_uri(repo, action="register"), "http://localhost/login/orcid/?action=register")
```

Comment:
This test is failing for me because it's missing some data around the request. But once that data is provided and the function runs, I think it will still not be a great idea to pass query parameters in the `path` argument of `site_url` and `build_url` functions. I understand why you are putting it in the path and not a query dict or something, because [we don't provide a way to pass query parameters](https://github.com/BirkbeckCTP/janeway/issues/4321) into that function and even if you could, the [URL encoding is redundant](https://github.com/BirkbeckCTP/janeway/issues/4291) so the URL would be garbled on its way back from ORCiD.

I discovered these things while working on a separate feature, #3899, whereby users will be carried through the login and registration process with a bit more care. It's still a work in progress but I'm thinking of expanding simplifying and unifying our token models so that we can retrieve next URLs and query parameters like this, even when we send things out to ORCiD.

If we were to fix up the tokens, we could make use of ORCiD's [`state` parameter](https://info.orcid.org/documentation/integration-guide/customizing-the-sign-in-register-screen/#h-identify-the-researcher-by-a-custom-state-parameter), which would let us link up the appropriate "next" url, whether login or registration (your feature) or something else (my feature).

I know this is a lot but I am curious what your thoughts are. If you want to glance over my still-messy WIP, it's here: https://github.com/BirkbeckCTP/janeway/pull/4322

Reply:
hi @joemull -- sorry for my delayed response.  I got waylaid by other tasks and some time off. I've checked in corrections to the more minor issues you've pointed out.

I like your idea for fixing the return url business.  I'm also thinking we might be able to use this if fix something else Alainna requested. She doesn't want to have to add each new journal/repo url to the redirect urls but instead have a press or installation wide return url that is able to redirect appropriately.

Reply:
@joemull How do you want to proceed with this work?  We'd like to get this and a few other orcid features that would depend on this rewrite into the next release.

Reply:
@everreau I think unifying the various `Token` models to a single one and passing the token ID in `state` to the ORCID API is the right direction. I could potentially separate that work from the rest of my branch and put it through review with higher priority. You could then rebase this branch on that and finish up this work.

I'd have to run it by @mauromsl and @ajrbyers to confirm they like the solution and figure out a time frame. It will be tight, because we are hoping to finish up 1.7 in the next few weeks.

Shall I do that and let you know?

Or if you have time to handle it ASAP, would you want to copy-paste and fix up my WIP code that unifies the Token objects and sends and receives `state` as a part of this PR?

Reply:
@joemull I can probably do it. I think this is my highest priority task at the moment. I'll check with Chad to be sure. Thank you!

Reply:
That would be super! Let me know if I can help with context.

Reply:
@everreau I just had a chat with Andy and Mauro, and can we actually ask you to hold off on this? We need a chance to discuss it properly.

Reply:
@joemull no problem.  Chad has indeed deemed this my most important task so we'd like to get going as soon as we can.  Thanks!